src/Future.ts

```diff
@@ -0,0 +1,62 @@
+import * as Async from "./Async";
+import * as Fx from "./Fx";
+
+export interface Future<E, A> {
+  readonly state: FutureState<E, A>;
+  readonly wait: Fx.Fx<Async.Async | E, A>;
+  readonly complete: (fx: Fx.Fx<E, A>) => boolean;
+}
+
+export type FutureState<E, A> = Resolving<E, A> | Resolved<E, A>;
+
+export class Resolving<E, A> {
+  readonly tag = "Resolving";
+  constructor(readonly observers: ((fx: Fx.Fx<E, A>) => void)[] = []) {}
+}
+
+export class Resolved<E, A> {
+  readonly tag = "Resolved";
+  constructor(readonly fx: Fx.Fx<E, A>) {}
+}
+
+export const make =<E, A>(): Fx.Fx<never, Future<E, A>> =>
```

Comment:
Any reason we can't use a class with a private `_state` instead of a closure?

Reply:
I don't think so 😀